assistant/assistant.go

```diff
@@ -0,0 +1,186 @@
+		astConfig.Path = "/echo"
+	}
+	if astConfig.RetryInterval == 0 {
+		astConfig.RetryInterval = 15
+	}
+}
+
+// NewAssistant returns an Assistant for the given Config.
+func NewAssistant(ctx context.Context, config *AssistantConfig, agent *agent.Agent) (*Assistant, error) {
+	config.fillDefaults()
+	var addr = flag.String("addr", config.Host, "http service address")
+	u := url.URL{Scheme: "ws", Host: *addr, Path: config.Path}
+
+	header := http.Header{}
+
+	if v, exists := os.LookupEnv("INFLUX_TOKEN"); exists {
+		header.Add("Authorization", "Token "+v)
+	} else {
+		return nil, fmt.Errorf("Influx authorization token not found, please set in env")
+	}
+
+	// creates a new websocket connection
+	log.Printf("D! [assistant] Attempting connection to [%s]", config.Host)
+	ws, _, err := websocket.DefaultDialer.Dial(u.String(), header)
+	for err != nil { // on error, retry connection again
+		log.Printf("E! [assistant] Failed to connect to [%s], retrying in %ds, "+
+			"error was '%s'", config.Host, config.RetryInterval, err)
+
+		err := internal.SleepContext(ctx, time.Duration(config.RetryInterval)*time.Second)
+		if err != nil {
+			return nil, err
+		}
+
+		ws, _, err = websocket.DefaultDialer.Dial(u.String(), header)
+	}
+	log.Printf("D! [assistant] Successfully connected to %s", config.Host)
+
+	a := &Assistant{
+		config:     config,
+		connection: ws,
+		ctx:        ctx,
+		done:       make(chan bool),
+		agent:      agent,
+	}
+
+	go a.listenToServer()
+
+	return a, nil
+}
+
+// Stop is used to clean up active connection and all channels
+func (assistant *Assistant) Stop() {
+	assistant.connection.Close()
+	assistant.done <- true
+	close(assistant.done)
+}
+
+type plugin struct {
+	Name   string
+	Type   string
+	Config map[string]interface{}
+}
+
+type Request struct {
+	Operation string
+	Uuid      string
+	Plugin    plugin
+}
+
+type Response struct {
+	Status string
+	Uuid   string
+	Data   interface{}
+}
+
+// WriteToServer is used as a helper function to write status responses to server.
+func (assistant *Assistant) WriteToServer(payload Response) error {
+	/*
+		TODO: Write a dedicated struct for the input of WriteToServer.
+		Refer to the design doc for implementation.
+		https://docs.google.com/document/d/1pnXrWgXCvlpe5tB3YAlOtO6rsvBxoV7DMPHL82KpCok/edit?usp=sharing
+	*/
+
+	err := assistant.connection.WriteJSON(payload)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (assistant *Assistant) checkForTermination() {
+	for {
+		select {
+		case <-assistant.done:
+			return
+		case <-assistant.ctx.Done():
+			assistant.Stop()
+		default:
+			continue
+		}
+	}
+}
+
+// listenToServer takes requests from the server and puts it in Requests.
+func (assistant *Assistant) listenToServer() {
+	go assistant.checkForTermination()
+
+	for {
+		var req Request
+		err := assistant.connection.ReadJSON(&req)
+		if err != nil {
+			fmt.Println(err)
+		}
+
+		var data string
+		var res Response
+		switch req.Operation {
```

Comment:
Is the plan to eventually get the response from each cases relevant handler function?

Like:
```
res = getPluginConfigHandler(req)
```
or something like that?

Reply:
I think this will depend on how we want to structure the functions in agent. My initial thought is that the agent functions will return some amount of data that the assistant will construct into a response so the agent doesn't have to worry about the actual response struct. Curious if you think there are benefits to using the method your code kinda follows

Reply:
I'd say if there's going to be a good amount of building of the response objects, then probably keep them out of this function, just because you want to modularize things. For that, you're right, I don't think the agent should be building the responses. Because for example, what if they want to use some other agent function like StopPlugin, but outside of the assistant, then it wouldn't make sense to respond with assistant specific info. So I think there should probably still be wrapping functions, however probably in the Assistant, not the agent. 

Honestly not sure though, it really depends on how much code would be in each case. So we can reevaluate once we're closer to integration

Reply:
Yeah I agree with what you said. Keeping things in the assistant but in a separate method seem like the way to go